### PR TITLE
Separate extraction and decompression logic in datasets.utils.extract_archive

### PR DESCRIPTION
**Pull request by pmeier**
Currently the implementation of `datasets.utils.extract_archive` has two downsides:

1. It is not extendible. If a new archive type or compression needs to be added (see https://github.com/pytorch/vision/pull/3420#discussion_r580123477) one needs to write a new `_is_whatever` method and implement the processing in the large `if` / `elif` statement.
  
    This can get confusing quite fast especially since the archive type and compression are orthogonal concepts. This leads to structures like

    https://github.com/pytorch/vision/blob/67b25288ca202d027e8b06e17111f1bcebd2046c/torchvision/datasets/utils.py#L250-L251

    Imagine we had another archive type that is also compressed with `gzip`.

2. It mixes the extraction and decompression concept. This leads to constructs like this

     https://github.com/pytorch/vision/blob/67b25288ca202d027e8b06e17111f1bcebd2046c/torchvision/datasets/utils.py#L271-L274

    Note that `to_path` would need to be fixed for every case where we only decompress the file.

---

This PR overcomes these by separating the extraction and decompression logic:

- To add a new archive type one has to provide an `_extract_whatever` function and add it to `_ARCHIVE_EXTRACTORS` with the corresponding extension.
- To add a compression one has to provide an `open_whatever` function and add it to `_COMPRESSED_FILE_OPENERS` with the corresponding extension.

This PR is fully BC, but makes it a lot easier to move forward in the future.

### CODE REVIEW

File: test/test_datasets_utils.py [hunk 1]
```diff
@@ -100,6 +102,93 @@ def test_download_url_dispatch_download_from_google_drive(self, mock):
...
+
+    def test_detect_file_type_unknown_compression(self):
+        with self.assertRaises(RuntimeError):
+            utils._detect_file_type("foo.tar.baz")
+
+    def test_detect_file_type_unknown_partial_ext(self):
+        with self.assertRaises(RuntimeError):
+            utils._detect_file_type("foo.bar")
+
+    def test_decopress_gzip(self):
```
Comment (fmassa) [inline test/test_datasets_utils.py:138]: nit:
```suggestion
    def test_decompress_gzip(self):
```

File: test/test_datasets_utils.py [hunk 2]
```diff
@@ -100,6 +102,93 @@ def test_download_url_dispatch_download_from_google_drive(self, mock):
...
+            compressed, file, content = create_compressed(temp_dir)
+
+            utils.decompress(compressed)
+
+            self.assertTrue(os.path.exists(file))
+
+            with open(file, "r") as fh:
+                self.assertEqual(fh.read(), content)
+
+    def test_decopress_lzma(self):
```
Comment (fmassa) [inline test/test_datasets_utils.py:158]:
```suggestion
    def test_decompress_lzma(self):
```

File: torchvision/datasets/utils.py [hunk 3]
```diff
@@ -240,56 +243,149 @@ def _save_response_content(
...
+
+    # check if the suffix is a compression
+    with contextlib.suppress(RuntimeError):
+        _verify_compression(suffix)
+        return suffix, None, suffix
+
+    raise RuntimeError(f"Suffix '{suffix}' is neither recognized as archive type nor as compression.")
+
+
+def decompress(from_path: str, to_path: Optional[str] = None, remove_finished: bool = False) -> str:
```
Comment (fmassa) [inline torchvision/datasets/utils.py:320]: I would encourage not using this API for the rest of torchvision, and just calling `extract_archive`, which also handles compressed file.

From this perspective, I would say that this is an internal function and shouldn't be exposed

**Review by fmassa — APPROVED** [review]
Approving to unblock, I have a couple of comments

File: torchvision/datasets/utils.py [hunk 4]
```diff
@@ -242,56 +245,145 @@ def _save_response_content(
...
+    suffix = path.suffix
+    suffixes = pathlib.Path(file).suffixes
+    if not suffixes:
+        raise RuntimeError(
+            f"File '{file}' has no suffixes that could be used to detect the archive type and compression."
+        )
+    elif len(suffixes) > 2:
+        raise RuntimeError(
+            "Archive type and compression detection only works for 1 or 2 suffixes. " f"Got {len(suffixes)} instead."
+        )
```
Comment (adamjstewart) [inline torchvision/datasets/utils.py:296]: This introduces a bug for downloads with a period in the file name. For example, https://landcover.ai/download/landcover.ai.v1.zip.

Comment (adamjstewart) [reply torchvision/datasets/utils.py:296]: Working on a PR to fix this, will ping you when it's done.

Comment (adamjstewart) [reply torchvision/datasets/utils.py:296]: See #4099

**Pull request merged**